File: tckit/adapters/docs_searchers/_infosys_parser.py
"""_infosys_parser — private HTML parsing utilities for infosys.beckhoff.com.

Uses BeautifulSoup to extract clean text, titles, and structured data
(parameter tables) from Beckhoff infosys pages.

Infosys HTML structure (as of 2024):
  - Main content is in <div id="content"> or <div class="topic">
  - Navigation/header/footer can be stripped by tag: <nav>, <header>, <footer>
  - Parameter tables: <table> with <th> headers like "Name", "Type", "Direction"
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from bs4 import BeautifulSoup, Tag

if TYPE_CHECKING:
    pass


def extract_title(soup: BeautifulSoup) -> str:
    """Extract page title from <title> tag, cleaned up."""
    title_tag = soup.find("title")
    if title_tag and title_tag.string:
        # Strip " - Beckhoff Infosys" suffix if present
        title = title_tag.string.strip()
        for suffix in (" - Beckhoff Infosys", " | Beckhoff", " - TwinCAT"):
            if title.endswith(suffix):
                title = title[: -len(suffix)].strip()
        return title
    return ""


def extract_main_content(soup: BeautifulSoup) -> str:
    """Extract the main text content, stripping nav/header/footer boilerplate.

    Returns clean text with whitespace normalised.
    """
    # Remove boilerplate elements
    for tag in soup.find_all(["nav", "header", "footer", "script", "style"]):
        tag.decompose()

    # Try known content containers first
    for selector in ("div#content", "div.topic", "main", "article", "div.content"):
        container = soup.select_one(selector)
        if container and isinstance(container, Tag):
            return _clean_text(container.get_text(separator="\n"))

    # Fall back to body text
    body = soup.find("body")
    if body and isinstance(body, Tag):
        return _clean_text(body.get_text(separator="\n"))

    return _clean_text(soup.get_text(separator="\n"))


def extract_description(soup: BeautifulSoup) -> str:
    """Extract the first meaningful paragraph as the FB/function description."""
    for selector in ("div#content p", "div.topic p", "main p", "p"):
        tag = soup.select_one(selector)
        if tag and isinstance(tag, Tag):
            text = tag.get_text(separator=" ").strip()
            if len(text) > 20:  # ignore trivial fragments
                return text
    return ""


def extract_parameter_table(soup: BeautifulSoup) -> list[dict[str, str]]:
    """Extract parameter rows from Beckhoff parameter tables.

    Returns a list of dicts with keys: name, type, direction, description.
    Handles two Beckhoff table layouts:
      1. Single table with a "Direction" column (older style)
      2. Separate tables per VAR_INPUT / VAR_OUTPUT block, identified by
         the nearest preceding heading (newer TF6xxx style)
    """
    rows: list[dict[str, str]] = []

    for table in soup.find_all("table"):
        if not isinstance(table, Tag):
            continue
        header_row = table.find("tr")
        if not header_row or not isinstance(header_row, Tag):
            continue

        headers = [
            th.get_text(strip=True).lower()
            for th in header_row.find_all(["th", "td"])
        ]

        # Only process tables that look like parameter tables
        if not any(h in headers for h in ("name", "variable", "parameter")):
            continue

        col = {
            "name": _find_col(headers, ("name", "variable", "parameter")),
            "type": _find_col(headers, ("type", "data type")),
            "direction": _find_col(headers, ("direction", "access", "i/o")),
            "description": _find_col(headers, ("description", "comment", "meaning")),
        }

        # Infer direction from preceding heading when there's no direction column
        heading_direction = ""
        if col["direction"] is None:
            heading_direction = _infer_direction_from_heading(table)

        for tr in table.find_all("tr")[1:]:  # skip header row
            if not isinstance(tr, Tag):
                continue
            cells = [td.get_text(strip=True) for td in tr.find_all(["td", "th"])]
            if not cells:
                continue
            direction = _cell(cells, col["direction"]) or heading_direction
            rows.append(
                {
                    "name": _cell(cells, col["name"]),
                    "type": _cell(cells, col["type"]),
                    "direction": direction,
                    "description": _cell(cells, col["description"]),
                }
            )

    return rows


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def parse_html(html: str) -> BeautifulSoup:
    return BeautifulSoup(html, "html.parser")


def _clean_text(text: str) -> str:
    """Collapse runs of whitespace/blank lines."""
    lines = [line.strip() for line in text.splitlines()]
    # Remove consecutive blank lines
    result: list[str] = []
    prev_blank = False
    for line in lines:
        is_blank = line == ""
        if is_blank and prev_blank:
            continue
        result.append(line)
        prev_blank = is_blank
    return "\n".join(result).strip()


_INPUT_HEADINGS = ("var_input", "input", "inputs", "varat_eingang", "eingang")
_OUTPUT_HEADINGS = ("var_output", "output", "outputs", "var_in_out", "in_out", "ausgang")


def _infer_direction_from_heading(table: Tag) -> str:
    """Walk backwards through siblings/parents to find a heading that implies direction.

    Beckhoff TF6xxx pages use separate tables per direction block, titled
    e.g. 'VAR_INPUT', 'Inputs', 'VAR_OUTPUT', 'Outputs'.
    """
    # Check previous siblings first (heading directly above the table)
    for sibling in table.find_previous_siblings():
        if not isinstance(sibling, Tag):
            continue
        if sibling.name in ("h1", "h2", "h3", "h4", "h5", "p", "caption"):
            text = sibling.get_text(strip=True).lower()
            if any(kw in text for kw in _INPUT_HEADINGS):
                return "input"
            if any(kw in text for kw in _OUTPUT_HEADINGS):
                return "output"
            break  # stop at first heading-like element

    # Check table caption
    caption = table.find("caption")
    if caption and isinstance(caption, Tag):
        text = caption.get_text(strip=True).lower()
        if any(kw in text for kw in _INPUT_HEADINGS):
            return "input"
        if any(kw in text for kw in _OUTPUT_HEADINGS):
            return "output"

    return ""


def _find_col(headers: list[str], candidates: tuple[str, ...]) -> int | None:
    for candidate in candidates:
        for i, h in enumerate(headers):
            if candidate in h:
                return i
    return None


def _cell(cells: list[str], col: int | None) -> str:
    if col is None or col >= len(cells):
        return ""
    return cells[col]

File: dotnet/src/TcKit.Adapters.Docs/InfosysParser.cs
using System.Text.RegularExpressions;
using AngleSharp.Dom;
using AngleSharp.Html.Dom;
using AngleSharp.Html.Parser;

namespace TcKit.Adapters.Docs;

/// <summary>
/// One parsed parameter-table row (name / type / direction / description), before the FB split into
/// inputs and outputs.
/// </summary>
internal sealed record ParamRow(string Name, string Type, string Direction, string Description);

/// <summary>One property/value row from a hardware "Technical data" table.</summary>
internal sealed record TechRow(string Property, string Value);

/// <summary>
/// HTML parsing for infosys.beckhoff.com pages: clean text, titles, the first descriptive paragraph,
/// and parameter tables. Mirrors the Python <c>_infosys_parser</c> module; AngleSharp's CSS selectors
/// stand in for BeautifulSoup's <c>select_one</c>/<c>find_all</c>.
///
/// Infosys HTML (as of 2024): main content sits in <c>div#content</c> or <c>div.topic</c>; parameter
/// tables carry <c>Name</c>/<c>Type</c>/<c>Direction</c> headers, either as one table with a
/// Direction column (older style) or one table per VAR_INPUT/VAR_OUTPUT block (newer TF6xxx style).
/// </summary>
internal static class InfosysParser
{
    private static readonly HtmlParser Parser = new();

    // A Beckhoff order number, e.g. EL3004, EPP1008-0001 — used to drop comparison-header rows.
    private static readonly Regex OrderNumberLike =
        new(@"^[A-Z]{2,4}\d{3,4}(-\d{3,4})?$", RegexOptions.Compiled);

    private static readonly string[] InputHeadings =
        ["var_input", "input", "inputs", "varat_eingang", "eingang"];

    private static readonly string[] OutputHeadings =
        ["var_output", "output", "outputs", "var_in_out", "in_out", "ausgang"];

    public static IDocument Parse(string html) => Parser.ParseDocument(html);

    /// <summary>Extract the page title from &lt;title&gt;, stripping Beckhoff branding suffixes.</summary>
    public static string ExtractTitle(IDocument doc)
    {
        var title = doc.QuerySelector("title")?.TextContent.Trim() ?? "";
        foreach (var suffix in new[] { " - Beckhoff Infosys", " | Beckhoff", " - TwinCAT" })
        {
            if (title.EndsWith(suffix, StringComparison.Ordinal))
            {
                title = title[..^suffix.Length].Trim();
            }
        }

        return title;
    }

    /// <summary>Extract the main text content, stripping nav/header/footer/script/style boilerplate.</summary>
    public static string ExtractMainContent(IDocument doc)
    {
        foreach (var node in doc.QuerySelectorAll("nav, header, footer, script, style"))
        {
            node.Remove();
        }

        foreach (var selector in new[] { "div#content", "div.topic", "main", "article", "div.content" })
        {
            var container = doc.QuerySelector(selector);
            if (container is not null)
            {
                return CleanText(NewlineText(container));
            }
        }

        var body = doc.Body;
        return body is not null ? CleanText(NewlineText(body)) : CleanText(NewlineText(doc.DocumentElement));
    }

    /// <summary>Extract the first meaningful paragraph as the FB/function description.</summary>
    public static string ExtractDescription(IDocument doc)
    {
        foreach (var selector in new[] { "div#content p", "div.topic p", "main p", "p" })
        {
            var tag = doc.QuerySelector(selector);
            if (tag is null)
            {
                continue;
            }

            var text = SpaceText(tag);
            if (text.Length > 20)
            {
                return text;
            }
        }

        return "";
    }

    /// <summary>
    /// Extract parameter rows from infosys parameter tables. Handles both the single-table layout
    /// (Direction column) and the per-direction-block layout (direction inferred from the nearest
    /// preceding heading or the table caption).
    /// </summary>
    public static IReadOnlyList<ParamRow> ExtractParameterTable(IDocument doc)
    {
        var rows = new List<ParamRow>();

        foreach (var table in doc.QuerySelectorAll("table"))
        {
            var headerRow = table.QuerySelector("tr");
            if (headerRow is null)
            {
                continue;
            }

            var headers = headerRow.QuerySelectorAll("th, td")
                .Select(c => StripText(c).ToLowerInvariant())
                .ToList();

            if (!headers.Any(h => h is "name" or "variable" or "parameter"))
            {
                continue;
            }

            var nameCol = FindCol(headers, "name", "variable", "parameter");
            var typeCol = FindCol(headers, "type", "data type");
            var directionCol = FindCol(headers, "direction", "access", "i/o");
            var descCol = FindCol(headers, "description", "comment", "meaning");

            var headingDirection = directionCol is null ? InferDirectionFromHeading(table) : "";

            foreach (var tr in table.QuerySelectorAll("tr").Skip(1))
            {
                var cells = tr.QuerySelectorAll("td, th").Select(StripText).ToList();
                if (cells.Count == 0)
                {
                    continue;
                }

                var direction = Cell(cells, directionCol);
                if (direction.Length == 0)
                {
                    direction = headingDirection;
                }

                rows.Add(new ParamRow(
                    Cell(cells, nameCol),
                    Cell(cells, typeCol),
                    direction,
                    Cell(cells, descCol)));
            }
        }

        return rows;
    }

    /// <summary>Read a &lt;meta name="..."&gt; content value, or null when absent.</summary>
    public static string? ExtractMeta(IDocument doc, string name)
    {
        var value = doc.QuerySelector($"meta[name={name}]")?.GetAttribute("content");
        return string.IsNullOrEmpty(value) ? null : value;
    }

    /// <summary>
    /// Return the href of the anchor whose visible text names the order number, or null. The anchor
    /// text may carry an exact order ("EL3004"), a variant-suffixed order ("EPP1008-0001"), or one-or-
    /// more orders in family-wildcard form ("EK110x-00xx, EK15xx", "EPI1008-000x, ERI1008-000x"). Each
    /// comma/space/slash-separated token is matched with the same x-wildcard and variant-suffix rules
    /// used to pick the section (<see cref="InfosysNavigator.SectionCoversOrder"/>). An exact
    /// (variant-suffix) match is preferred over a wildcard one, so a specific "EP3174-0002" entry wins
    /// over a broad "EP31xx-xxxx" group heading on the same overview. Whitespace-normalised,
    /// case-insensitive.
    /// </summary>
    public static string? FindLinkByOrder(IDocument doc, string order)
    {
        string? wildcardHref = null;
        foreach (var a in doc.QuerySelectorAll("a[href]"))
        {
            var text = StripText(a);
            foreach (var token in text.Split([',', ' ', '/'], StringSplitOptions.RemoveEmptyEntries))
            {
                // Exact: the token's base (variant suffix dropped) equals the order, e.g.
                // "EP3174-0002" -> "EP3174". Returns at once.
                if (token.Split('-')[0].Equals(order, StringComparison.OrdinalIgnoreCase))
                {
                    return a.GetAttribute("href");
                }

                // Wildcard / group: e.g. "EK110x-00xx" covers EK1100. Kept only if no exact is found.
                if (wildcardHref is null && InfosysNavigator.SectionCoversOrder(token, order))
                {
                    wildcardHref = a.GetAttribute("href");
                }
            }
        }

        return wildcardHref;
    }

    /// <summary>
    /// Extract a hardware "Technical data" table as property/value rows. Each row contributes its last
    /// two non-empty cells (so a 2-column EL table "property | value" and a 3-column EtherCAT Box table
    /// "category | property | value" both yield property/value). A table explicitly marked "Technical
    /// data" (first row, caption, or a preceding heading) is preferred; otherwise the richest
    /// property/value table on the page is used, which suits the dedicated technical-data page
    /// find_hardware navigates to. Returns an empty list when nothing table-like is found.
    /// </summary>
    public static IReadOnlyList<TechRow> ExtractTechnicalData(IDocument doc)
    {
        var tables = doc.QuerySelectorAll("table");

        foreach (var table in tables)
        {
            if (IsMarkedTechnicalData(table))
            {
                var marked = ParseKeyValueRows(table);
                if (marked.Count > 0)
                {
                    return marked;
                }
            }
        }

        var best = new List<TechRow>();
        foreach (var table in tables)
        {
            var rows = ParseKeyValueRows(table);
            if (rows.Count > best.Count)
            {
                best = rows;
            }
        }

        // A handful of rows guards against picking up an incidental two-cell table.
        return best.Count >= 3 ? best : [];
    }

    private static List<TechRow> ParseKeyValueRows(IElement table)
    {
        var rows = new List<TechRow>();
        foreach (var tr in table.QuerySelectorAll("tr"))
        {
            var cells = tr.QuerySelectorAll("td, th").Select(StripText).Where(c => c.Length > 0).ToList();
            if (cells.Count < 2)
            {
                continue;
            }

            var property = cells[^2];
            if (property.Equals("technical data", StringComparison.OrdinalIgnoreCase)
                || OrderNumberLike.IsMatch(property))
            {
                // The "Technical data | <order>" header row, or an EtherCAT Box comparison header
                // whose cells are the variant order numbers (e.g. EPP1008-0001 | EPP1018-0001).
                continue;
            }

            rows.Add(new TechRow(property, cells[^1]));
        }

        return rows;
    }

    private static bool IsMarkedTechnicalData(IElement table)
    {
        var firstRow = table.QuerySelector("tr");
        if (firstRow is not null
            && string.Join(" ", firstRow.QuerySelectorAll("td, th").Select(StripText))
                .Contains("technical data", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        var caption = table.QuerySelector("caption");
        if (caption is not null && StripText(caption).Contains("technical data", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        for (var sib = table.PreviousElementSibling; sib is not null; sib = sib.PreviousElementSibling)
        {
            if (sib.TagName is "H1" or "H2" or "H3" or "H4" or "H5" or "H6" or "P" or "CAPTION" or "STRONG" or "B")
            {
                return StripText(sib).Contains("technical data", StringComparison.OrdinalIgnoreCase);
            }
        }

        return false;
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    /// <summary>Text content with a newline between every text node (BeautifulSoup get_text("\n")).</summary>
    private static string NewlineText(INode? node) =>
        node is null ? "" : string.Join("\n", node.Descendants<IText>().Select(t => t.Text));

    /// <summary>Whitespace-normalised text content (BeautifulSoup get_text(" ").strip()).</summary>
    private static string SpaceText(IElement el) => Regex.Replace(el.TextContent, @"\s+", " ").Trim();

    /// <summary>Whitespace-normalised text content for a single cell/title (BeautifulSoup strip=True).</summary>
    private static string StripText(IElement el) => Regex.Replace(el.TextContent, @"\s+", " ").Trim();

    private static string CleanText(string text)
    {
        var lines = text.Split('\n').Select(l => l.Trim());
        var result = new List<string>();
        var prevBlank = false;
        foreach (var line in lines)
        {
            var isBlank = line.Length == 0;
            if (isBlank && prevBlank)
            {
                continue;
            }

            result.Add(line);
            prevBlank = isBlank;
        }

        return string.Join("\n", result).Trim();
    }

    private static string InferDirectionFromHeading(IElement table)
    {
        for (var sibling = table.PreviousElementSibling; sibling is not null; sibling = sibling.PreviousElementSibling)
        {
            if (sibling.TagName is "H1" or "H2" or "H3" or "H4" or "H5" or "P" or "CAPTION")
            {
                var text = StripText(sibling).ToLowerInvariant();
                if (InputHeadings.Any(text.Contains))
                {
                    return "input";
                }

                if (OutputHeadings.Any(text.Contains))
                {
                    return "output";
                }

                break; // stop at the first heading-like element
            }
        }

        var caption = table.QuerySelector("caption");
        if (caption is not null)
        {
            var text = StripText(caption).ToLowerInvariant();
            if (InputHeadings.Any(text.Contains))
            {
                return "input";
            }

            if (OutputHeadings.Any(text.Contains))
            {
                return "output";
            }
        }

        return "";
    }

    private static int? FindCol(List<string> headers, params string[] candidates)
    {
        foreach (var candidate in candidates)
        {
            for (var i = 0; i < headers.Count; i++)
            {
                if (headers[i].Contains(candidate, StringComparison.Ordinal))
                {
                    return i;
                }
            }
        }

        return null;
    }

    private static string Cell(List<string> cells, int? col) =>
        col is null || col >= cells.Count ? "" : cells[col.Value];
}
